### PR TITLE
basichost: ensure no duplicates in Addrs output

### PR DESCRIPTION
**Pull request by sukunrt**
fixes: https://github.com/libp2p/go-libp2p/issues/2972

### CODE REVIEW

**Comment by Wondertan** [comment]
I can help testing this on my cluster once the fix is final

**Comment by sukunrt** [comment]
@Wondertan, The fix is final. Can you check this?

**Comment by Wondertan** [comment]
Looks good!
image

**Comment by Wondertan** [comment]
Might be worth adding a unit test

File: p2p/host/basic/basic_host.go [hunk 1]
```diff
@@ -500,13 +500,17 @@ func (h *BasicHost) makeUpdatedAddrEvent(prev, current []ma.Multiaddr) *event.Ev
 		return nil
 	}
 	prevmap := make(map[string]ma.Multiaddr, len(prev))
+	currmap := make(map[string]ma.Multiaddr, len(current))
```
Comment (MarcoPolo) [inline p2p/host/basic/basic_host.go:503]: Do we even use `currmap`?

**Review by MarcoPolo — APPROVED** [review]
Looks good minus the superfluous `currmap`

File: p2p/host/basic/basic_host.go (hunk 1 above)
Comment (sukunrt) [reply p2p/host/basic/basic_host.go:503]: I changed the iteration below to be over this map. This ensures that this method works even if there are duplicates in `Addrs` output.

**Comment by MarcoPolo** [comment]
@dimartiro Yes. Removing duplicates without allocating means that you have to change the order of the returned Addrs (https://github.com/multiformats/go-multiaddr/blob/d19cf5da7dc590c829b9a4b329341ea8847d4d2a/multiaddr.go#L253).

If you don't have duplicates, it shouldn't affect the order.

**Pull request merged**